Removing allocation from PathFinder.search and Room.find calls

daboross opened the pull request:
This is slightly more unsafe code, but should still be sound. The principal is the same as transmuting `'a` to `'static` - it just affects the reference to the `Fn` as well now, not just the lifetime of the data within the box.

The actual change is that instead of allocating a space for the callback on the heap and storing a pointer to it, we just store a pointer to the stack location it's in within the pathfinder function. This change is made for both `Room.find` and `PathFinder.search`.

ASalvail commented on screeps-game-api/src/objects/impls/room.rs:
```diff
@@ -170,18 +170,17 @@ impl Room {
+        let callback_type_erased: &(Fn(String, Reference) -> Option<Reference> + 'a) =
+            &callback_boxed;
+
+        // Overwrite lifetime of box inside closure so it can be stuck in scoped_thread_local
```
Still refers to the `Box`

ASalvail commented on screeps-game-api/src/pathfinder.rs:
```diff
@@ -360,15 +360,15 @@ where
+    // now unified as &Fn
+    let callback_type_erased: &(Fn(String) -> Reference + 'a) = &callback_unboxed;
 
     // Overwrite lifetime of box inside closure so it can be stuck in scoped_thread_local storage:
```
Same with `Box`.

daboross replied:
These should be fixed!

daboross commented:
Rebased on master for changelog conflicts. Maybe not including those in PRs would be a good idea in the future, just for the sake of not conflicting on editing the same file... But it's nice having it already there? Not sure.

I think I've fixed all the box comments.
